## Decision Trees

## Import Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

## Load in Data

In [2]:
iris = sns.load_dataset('iris')

In [3]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

## Data Wrangling

### Specify x & y variables using subsetting

In [4]:
x = iris.drop('species', axis=1)
y = iris['species']

## Train Test Split

In [5]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=76)

## Create Initial Decision Tree

In [6]:
decisionTree = DecisionTreeClassifier(random_state=76)
decisionTree.fit(x_train, y_train)

DecisionTreeClassifier(random_state=76)

## Assess the Model

In [10]:
treePredictions = decisionTree.predict(x_test)
print(confusion_matrix(y_test, treePredictions))

[[19  0  0]
 [ 0 10  3]
 [ 0  2 11]]


## Reading the Confusion Matrix

### This decision tree model is really good at predicting the species of setosa, but misclassified a few of the versicolor species as virginica and vice versa.

## How Well Does the Model Fit?

In [11]:
print(classification_report(y_test, treePredictions))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        19
  versicolor       0.83      0.77      0.80        13
   virginica       0.79      0.85      0.81        13

    accuracy                           0.89        45
   macro avg       0.87      0.87      0.87        45
weighted avg       0.89      0.89      0.89        45



### Looking at precision: 100% for setosa, % for versicolor, and 79% for virginica, with 89% overall accuracy (weighted average).